url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/tree/XTreeMakeEasy.java

```diff
@@ -0,0 +1,191 @@
+
+import xscript.compiler.XDiagnostic;
+import xscript.compiler.XOperator;
+import xscript.compiler.tree.XTree.XTreeConstant;
+import xscript.compiler.tree.XTree.XTreeFor;
+import xscript.compiler.tree.XTree.XTreeGroup;
+import xscript.compiler.tree.XTree.XTreeIf;
+import xscript.compiler.tree.XTree.XTreeOperator;
+import xscript.compiler.tree.XTree.XTreeOperatorIf;
+import xscript.compiler.tree.XTree.XTreeOperatorPrefixSuffix;
+import xscript.compiler.tree.XTree.XTreeWhile;
+
+public class XTreeMakeEasy extends XTreeChanger {
+
+	private DiagnosticListener<String> diagnosticListener;
+	
+	private boolean doreplace;
+	
+	private XTree replace;
+	
+	public XTreeMakeEasy(DiagnosticListener<String> diagnosticListener){
+		this.diagnosticListener = diagnosticListener;
+	}
+	
+	@SuppressWarnings("unchecked")
+	@Override
+	public <T extends XTree> T visitTree(T tree){
+		if(tree!=null){
+			try{
+				tree.accept(this);
+			}catch(UnsupportedOperationException e){
+			}catch(Exception e){
+				addDiagnostic(tree, "makeeasy.tree.failure", e.getMessage());
+			}
+		}
+		if(doreplace){
+			tree = (T) replace;
+			doreplace = false;
+		}
+		return tree;
+	}
+	
+	@Override
+	public void visitOperator(XTreeOperator operator) {
+		super.visitOperator(operator);
+		XTree t = operator.exprs.get(0);
+		if(!(t instanceof XTreeConstant))
+			return;
+		XTreeConstant left = (XTreeConstant)t;
+		while(!operator.operators.isEmpty()){
+			XOperator o = operator.operators.get(0);
+			t = operator.exprs.get(1);
+			if(!(t instanceof XTreeConstant)){
+				break;
+			}
+			XTreeConstant right = (XTreeConstant)t;
+			try{
+				right.value = o.calc(left.value, right.value);
+			}catch(UnsupportedOperationException e){
+				break;
+			}catch(ClassCastException e){
+				addDiagnostic(right, "invalid.data.types", e.getMessage());
+				break;
+			}
+			left = right;
+			operator.operators.remove(0);
+			operator.exprs.remove(0);
+		}
+		if(operator.operators.isEmpty()){
+			setReplace(left);
+		}
+	}
+
+	@Override
+	public void visitOperatorPrefixSuffix(XTreeOperatorPrefixSuffix xOperatorPrefixSuffix) {
+		super.visitOperatorPrefixSuffix(xOperatorPrefixSuffix);
+		if(xOperatorPrefixSuffix.expr instanceof XTreeConstant){
+			Object value = ((XTreeConstant)xOperatorPrefixSuffix.expr).value;
+			if(xOperatorPrefixSuffix.prefix!=null){
+				//TODO also implement half resolve
+				for(XOperator op: xOperatorPrefixSuffix.prefix){
+					value = op.calc(value, null);
+				}
+			}
+			if(xOperatorPrefixSuffix.suffix!=null){
+				for(XOperator op: xOperatorPrefixSuffix.suffix){
+					value = op.calc(value, null);
+				}
+			}
+			replaceWith(xOperatorPrefixSuffix, value);
+		}
+	}
+	
+	@Override
+	public void visitWhile(XTreeWhile xWhile) {
+		super.visitWhile(xWhile);
+		if(xWhile.condition instanceof XTreeConstant){
+			Object value = ((XTreeConstant)xWhile.condition).value;
+			if(!nonZero(value)){
+				addDiagnostic(xWhile.body, "code.dead");
+				setReplace(null);
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 100:106, 112:118.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.